### PR TITLE
Add a SPIFFE based authenticator

### PR DESCRIPTION
**Pull request by hug-dev**
Uses the spiffe crate to connect to the SPIFFE Workload API in order to
verify JWT SVID and extract the SPIFFE ID to be used as application
name.

*Note:* after this PR is merged, the `spiffe` branch will be deleted.

### CODE REVIEW

File: src/authenticators/jwt_svid_authenticator/mod.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            id: AuthType::JwtSvid,
+        })
+    }
+
+    fn authenticate(
+        &self,
+        auth: &RequestAuth,
+        _: Option<ConnectionMetadata>,
+    ) -> Result<Application> {
+        let svid = str::from_utf8(auth.buffer.expose_secret()).map_err(|e| {
```
Comment (ionut-arm) [inline src/authenticators/jwt_svid_authenticator/mod.rs:58]: Does `svid` need zeroizing when you're done with it, or do you just get a `&str` here? (might've asked this in the previous PRs, actually)

**Review by ionut-arm — COMMENTED** [review]
Seems good, just a few comments

File: config.toml [hunk 2]
```diff
@@ -74,6 +74,12 @@ auth_type = "UnixPeerCredentials"
 # Read more here: https://parallaxsecond.github.io/parsec-book/parsec_client/operations/index.html#core-operations
 #admins = [ { name = "admin_1" }, { name = "admin_2" } ]
 
+# (Required only for JwtSvid) Location of the Workload API endpoint
+# WARNING: only use this authenticator if the Workload API socket is TRUSTED. A malicious entity
+# owning that socket would have access to all the keys owned by clients using this authentication
+# method. This path *must* be trusted for as long as Parsec is running.
+#workload_endpoint="unix:///tmp/agent.sock"
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply config.toml:81]: That is the [default](https://spiffe.io/docs/latest/deploying/spire_agent/#agent-configuration-file) unfortunately 😢
In [another guide](https://spiffe.io/docs/latest/deploying/install-agents/#step-3-create-agent-daemonset), they say `/run/spire/sockets`. So I can use `/run/spire/sockets/agent.sock` to be on the safe side.

File: e2e_tests/tests/per_provider/normal_tests/generate_random.rs [hunk 3]
```diff
@@ -53,5 +53,5 @@ fn generate_zero_bytes() {
     const NBYTES: usize = 0;
     let random_bytes = client.generate_bytes(NBYTES).unwrap();
 
-    assert_eq!(random_bytes, vec![]);
+    assert_eq!(random_bytes, Vec::<u8>::new());
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply e2e_tests/tests/per_provider/normal_tests/generate_random.rs:56]: Yes there was a compilation error :O

File: src/authenticators/jwt_svid_authenticator/mod.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Contributors to the Parsec project.
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply src/authenticators/jwt_svid_authenticator/mod.rs:1]: This was from the `spiffe` branch which was upstreamed last year so I think technically it's fine to keep it 2020

File: src/authenticators/jwt_svid_authenticator/mod.rs (hunk 1 above)
Comment (hug-dev) [reply src/authenticators/jwt_svid_authenticator/mod.rs:58]: In the new crate `validate_jwt_token` just need a `&str` so here I removed the `to_string`, it's only a `&str` everywhere so no need to zeroize 👌

**Review by ionut-arm — APPROVED** [review]

**Review by ionut-arm — APPROVED** [review]

**Pull request merged**